src/libs/ApiUtils.js

```diff
@@ -0,0 +1,78 @@
+import lodashGet from 'lodash/get';
+import Onyx from 'react-native-onyx';
+import ONYXKEYS from '../ONYXKEYS';
+import CONFIG from '../CONFIG';
+import CONST from '../CONST';
+import * as Environment from './Environment/Environment';
+
+// To avoid rebuilding native apps, native apps use production config for both staging and prod
+// We use the async environment check because it works on all platforms
+let ENV_NAME = CONST.ENVIRONMENT.PRODUCTION;
+Environment.getEnvironment()
+    .then((envName) => {
+        ENV_NAME = envName;
+    });
+
+let stagingServerToggleState;
+Onyx.connect({
+    key: ONYXKEYS.USER,
+    callback: (val) => {
+        stagingServerToggleState = lodashGet(val, 'shouldUseStagingServer');
+    },
+});
+
+function shouldUseStagingServer() {
+    // Choosing between staging and prod/dev is only available on DEV and STAGING
+    if (ENV_NAME === CONST.ENVIRONMENT.PRODUCTION) {
+        return false;
+    }
+
+    if (typeof stagingServerToggleState !== 'boolean') {
+        return ENV_NAME === CONST.ENVIRONMENT.STAGING;
```

Comment:
Could we make the default value for `stagingServerToggleState` be this comparison? Then simplify this method to:

```js
function shouldUseStagingServer() {
    return ENV_NAME === CONST.ENVIRONMENT.PRODUCTION 
        ? false 
        : stagingServerToggleState;
}
```

Reply:
Yes, though see this
https://github.com/Expensify/App/blob/bb0b7883210a9aba510adf682363004d536e29b4/src/libs/ApiUtils.js#L10-L16

we want the default value for staging to be `true`, but we can't simply do this
```js
let stagingServerToggleState = ENV_NAME === CONST.ENVIRONMENT.STAGING;
```

because the real ENV_NAME is resolved a few moments later

We can only do this:

```js
let stagingServerToggleState = false;
Onyx.connect({
    key: ONYXKEYS.USER,
    callback: (val) => {
        const defaultValue = ENV_NAME === CONST.ENVIRONMENT.STAGING;
        stagingServerToggleState = lodashGet(val, 'shouldUseStagingServer', defaultValue);
    },
});
```

It's probably fine, but ENV_NAME is obtained async and it might not be updated in the Onyx.connect callback 
IMO the safest is what we have right now
What do you think?

Reply:
>ENV_NAME is obtained async and it might not be updated in the Onyx.connect callback

Does it lead to any clear problem though? We can test to see if it's fine or not?

Reply:
> > ENV_NAME is obtained async and it might not be updated in the Onyx.connect callback
> 
> Does it lead to any clear problem though?

A risk STAGING might use the PROD API

I think it's a known risk when we [depend on 2 async callbacks where order matters](https://github.com/Expensify/App/blob/698ef4a97ee93a5db4014aca829a7fde4cca7cc7/src/libs/ApiUtils.js#L10-L35) 
- Onyx callback
- Env callack 

"Onyx callback" uses env value and depends that "Env callback" has already ran

The logic can only work correctly if "Onyx callback" runs after "Env Callback", but nothing ensures such order

Is it OK to declare like:
```js
// To avoid rebuilding native apps, native apps use production config for both staging and prod
// We use the async environment check because it works on all platforms
let ENV_NAME = CONST.ENVIRONMENT.PRODUCTION;
let stagingServerToggleState = false;

Environment.getEnvironment()
    .then((envName) => {
        ENV_NAME = envName;

        Onyx.connect({
          key: ONYXKEYS.USER,
          callback: (val) => {
              const defaultValue = ENV_NAME === CONST.ENVIRONMENT.STAGING;
              stagingServerToggleState = lodashGet(val, 'shouldUseStagingServer', defaultValue);
          },
      });
    });
```

Reply:
I've updated the code to subscribe to Onyx after `getEnvironment` resolves (like above sample) to avoid the risk 
Let's tie this thread and continue here if we need to: https://github.com/Expensify/App/pull/15178#discussion_r1132173422